### ISSUE TITLE
Load Nested field with only=str gives error when sent flat data structure instead of dictionary

### ISSUE THREAD

**Issue by erlingbo** [marshmallow-code/marshmallow#307]
When using a schema with the only attribute set to a string, the load method returns UnmarshalResult without errors nor data.

```
class MySchema(Schema):
    class Meta:
        fields = ('pk',)

    pk = fields.String()

MySchema(only='pk').load('an_id1234')
MySchema(only='pk').load({'pk': 'an_id1234'})
```

Both return the same UnmarshalResult
`UnmarshalResult(data={}, errors={})`

Preferably the first should work, but the second could be used with a pre load method on the schema class.

Is there anything I have done wrong or is this a bug?

Note: I use the same schema for dumping data an thus use `only='pk'` and not `only=('pk',)` for getting the correct dump output

**Comment by franciscod** [comment]
Since `'pk'` may be interpreted as `('p', 'k')`, you should use `('pk',)`, right?

When using only=('pk',) it gives a more interesting result:

``` python
>>> MySchema(only=('pk',)).load('an_id1234')
UnmarshalResult(data={}, errors={'_schema': ['Invalid input type.']})
>>> MySchema(only=('pk',)).load({'pk': 'an_id1234'})
UnmarshalResult(data={'pk': 'an_id1234'}, errors={})
```

https://marshmallow.readthedocs.org/en/latest/api_reference.html#schema
The `only` attribute expects a tuple or a list, if you give it a string it will interpret it as an iterable of chars:

``` python
>>> class CharSchema(Schema):
...     class Meta:
...             fields = ('a', 'b', 'c', 'd')
...     a = fields.String()
...     b = fields.String()
...     c = fields.String()
...     d = fields.String()
... 
>>> CharSchema(only='ab').load({'a': 'asd', 'b': 'qwe'})
UnmarshalResult(data={'b': 'qwe', 'a': 'asd'}, errors={})

```

**Comment by erlingbo** [comment]
My bad. I went to far when I tried to dumb down my problem. 

The real scenario is like this:

```
        class MyNestedSchema(Schema):
        class Meta:
            fields = ('pk',)

        pk = fields.String()

    class MySchema(Schema):
        class Meta:
            fields = ('pk', 'child')
        child = fields.Nested(MyNestedSchema, only='pk')

    class Obj(object):
        def __init__(self, pk, child):
            self.pk = pk
            self.child = child

    result = MySchema().dump(Obj(123, Obj(456, None)))
    print result
    print MySchema().load(result.data)
    print MySchema().load({'pk': 'an_id1234', 'child': 'child123'})
    print MySchema().load({'pk': 'an_id1234', 'child': {'pk': 'child123'}})
```

I use nested fields with only parameter as described in the [docs](http://marshmallow.readthedocs.org/en/latest/nesting.html#specifying-which-fields-to-nest). 

This gives me the following results:

```
MarshalResult(data={'pk': 123, 'child': u'456'}, errors={})
UnmarshalResult(data={'pk': 123}, errors={'child': {u'_schema': [u'Invalid input type.']}})
UnmarshalResult(data={'pk': 'an_id1234'}, errors={'child': {u'_schema': [u'Invalid input type.']}})
UnmarshalResult(data={'pk': 'an_id1234', 'child': {'pk': u'child123'}}, errors={})
```

Shouldn't the Nested field be able to handle this?

**Comment by franciscod** [comment]
``` python

    class MySchema(Schema):
        class Meta:
            fields = ('pk', 'child')
        child = fields.Nested(MyNestedSchema, only='pk')
```

Have you tried using

``` python
only=('pk',)
```

as opposed to 

``` python
only='pk'
```

?

**Comment by erlingbo** [comment]
Yes, that gives me a dump with a dictionary instead of a flat object, as expected

```
{'pk': '456'}
```

```
'456'
```

**Comment by franciscod** [comment]
I don't get what's the issue :disappointed:

**Comment by erlingbo** [comment]
Okei. I'll try to explain. I'll want a dump that gives me a file (json or yaml) and then be able to load the file again later. 

I want the relations to be flat, just using the primary key of the object it links to. This feature is documented [here](http://marshmallow.readthedocs.org/en/latest/nesting.html#specifying-which-fields-to-nest) as `only='pk'`

When I try to load the file using the same schema I get an error `UnmarshalResult(data={'pk': 123}, errors={'child': {u'_schema': [u'Invalid input type.']}})`

If I use `only=('pk',)` the dumped relations will not be flat and include an object with the `pk` as attribute. E.g. `{'pk': 123, 'child': {'pk': u'456'}}` vs what I want: ``{'pk': 123, 'child': u'456'}`

I need to integrate with other teams, and the files need to be as human readable as possible, thus I want the flat structure. When this is a documented feature of the Nested field, I also expect it to be able to load the corresponding data.

**Comment by sloria** [comment]
This is the same issue addressed in #316. Further discussion should happen there.
